### PR TITLE
Extractor uploads local Cromwell checkout and config to GCS [BA-6382]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 1]
```diff
@@ -22,40 +23,62 @@
...
-    #   git clone . --local "${git_snapshot}"
-    raise Exception("Not Implemented")
+def upload_local_checkout(cromwell_path: Path,
+                          gcs_bucket: str,
+                          gcs_path: str,
+                          gcs_storage_client: storage.Client) -> None:
+    cromwell_snapshots_path = cromwell_path.parent / "cromwell_snapshots"
+
+    repo = git.Repo(cromwell_path)
+    if repo.is_dirty():
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:67]: 🆒

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 2]
```diff
@@ -22,40 +23,62 @@
...
+        raise Exception("Unable to upload local checkout to GCS: repository is dirty - need to do check in first.")
+
+    last_commit_hash = repo.head.commit.hexsha
+    zip_file_name = f"{last_commit_hash}.zip"
+    zip_file_path = Path(cromwell_snapshots_path / zip_file_name)
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:72]:
```suggestion
    zip_file_path = cromwell_snapshots_path / zip_file_name
```

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 3]
```diff
@@ -106,9 +128,16 @@ def process_workflow(cromwell_url: str, gcs_bucket: str, gcs_path: str, gcs_stor
...
+    parser.add_argument('cromwell_url', metavar='CROMWELL', type=url_regex_validator, nargs=1,
+                        help='Cromwell host')
+    parser.add_argument('gcs_path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1,
+                        help='GCS path to upload to')
+    parser.add_argument('workflows', metavar='WORKFLOW', type=workflow_regex_validator, nargs='+',
+                        help='Workflows to process')
+    parser.add_argument('cromwell_checkout_path', metavar='CROMWELLCHECKOUTPATH', type=Path,
+                        help='Path to Cromwell git checkout used to run workflows')
+    parser.add_argument('cromwell_config_path', metavar='CROMWELLCONFIGPATH', type=Path,
+                        help='Path to Cromwell configuration file used to run workflows')
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:140]: why are these two new ones optional?

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 4]
```diff
@@ -22,40 +23,62 @@
...
+    zip_file_path = Path(cromwell_snapshots_path / zip_file_name)
+    if not os.path.exists(zip_file_path):
+        current_snapshot_path = __create_snapshot_of_local_repo(repo, cromwell_snapshots_path)
+        __create_zip_file(zip_file_path, current_snapshot_path)
+
+    upload_blob(gcs_bucket, zip_file_path.read_text(), f"{gcs_path}/{zip_file_name}", gcs_storage_client, logger)
+
+
+def upload_local_config(config_path: Path, gcs_bucket: str, gcs_path: str, gcs_storage_client: storage.Client):
+    upload_blob(gcs_bucket, config_path.read_text(), f"{gcs_path}/{config_path.name}", gcs_storage_client, logger)
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:81]: Same comment on standardized naming here as in the code snapshot.

File: scripts/metadata_comparison/metadata_comparison/extractor.py (hunk 2 above)
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:72]: I believe the snapshots path would be under the workflow id so I think there should be only one snapshot per workflow. If so a standardized name for the snapshot might be helpful for any downstream code that wants to consume it.

Comment (grsterin) [reply scripts/metadata_comparison/metadata_comparison/extractor.py]: Actually such change results in a warning on next line:
```
Unexpected type(s): (Path) Possible types: (Union[bytes, str, PathLike]) (Union[bytes, str, PathLike, int]) 
```

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 5]
```diff
@@ -106,9 +128,16 @@ def process_workflow(cromwell_url: str, gcs_bucket: str, gcs_path: str, gcs_stor
...
-    parser.add_argument('cromwell_url', metavar='CROMWELL', type=url_regex_validator, nargs=1, help='Cromwell host')
-    parser.add_argument('gcs_path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1, help='GCS path to upload to')
-    parser.add_argument('workflows', metavar='WORKFLOW', type=workflow_regex_validator, nargs='+', help='Workflows to process')
+    parser.add_argument('cromwell_url', metavar='CROMWELL', type=url_regex_validator, nargs=1,
+                        help='Cromwell host')
+    parser.add_argument('gcs_path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1,
+                        help='GCS path to upload to')
+    parser.add_argument('workflows', metavar='WORKFLOW', type=workflow_regex_validator, nargs='+',
+                        help='Workflows to process')
+    parser.add_argument('cromwell_checkout_path', metavar='CROMWELLCHECKOUTPATH', type=Path,
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:137]:
```suggestion
    parser.add_argument('cromwell-checkout-path', metavar='CROMWELLCHECKOUTPATH', type=Path,
```

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 6]
```diff
@@ -106,9 +128,16 @@ def process_workflow(cromwell_url: str, gcs_bucket: str, gcs_path: str, gcs_stor
     parser = argparse.ArgumentParser(
         description='Extract metadata and operation details for workflows and upload to GCS')
     parser.add_argument('-v', '--verbose', action='store_true')
-    parser.add_argument('cromwell_url', metavar='CROMWELL', type=url_regex_validator, nargs=1, help='Cromwell host')
-    parser.add_argument('gcs_path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1, help='GCS path to upload to')
-    parser.add_argument('workflows', metavar='WORKFLOW', type=workflow_regex_validator, nargs='+', help='Workflows to process')
+    parser.add_argument('cromwell_url', metavar='CROMWELL', type=url_regex_validator, nargs=1,
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:131]:
```suggestion
    parser.add_argument('cromwell-url', metavar='CROMWELL', type=url_regex_validator, nargs=1,
```

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 7]
```diff
@@ -106,9 +128,16 @@ def process_workflow(cromwell_url: str, gcs_bucket: str, gcs_path: str, gcs_stor
...
+                        help='Cromwell host')
+    parser.add_argument('gcs_path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1,
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:133]:
```suggestion
    parser.add_argument('gcs-path', metavar='GCSPATH', type=gcs_path_regex_validator, nargs=1,
```

File: scripts/metadata_comparison/metadata_comparison/extractor.py [hunk 8]
```diff
@@ -106,9 +128,16 @@ def process_workflow(cromwell_url: str, gcs_bucket: str, gcs_path: str, gcs_stor
...
+                        help='GCS path to upload to')
+    parser.add_argument('workflows', metavar='WORKFLOW', type=workflow_regex_validator, nargs='+',
+                        help='Workflows to process')
+    parser.add_argument('cromwell_checkout_path', metavar='CROMWELLCHECKOUTPATH', type=Path,
+                        help='Path to Cromwell git checkout used to run workflows')
+    parser.add_argument('cromwell_config_path', metavar='CROMWELLCONFIGPATH', type=Path,
```
Comment (mcovarr) [inline scripts/metadata_comparison/metadata_comparison/extractor.py:139]:
```suggestion
    parser.add_argument('cromwell-config-path', metavar='CROMWELLCONFIGPATH', type=Path,
```

**Comment by cjllanwarne** [comment]
TOL: You might want to add the `src/ci/bin/testMetadataComparisonPython.sh` file to the "scripts only" filter so that it only needs to run scripts tests if that file is updated

**Comment by cjllanwarne** [comment]
cf: https://github.com/broadinstitute/cromwell/blob/develop/src/ci/bin/test.inc.sh#L145

File: scripts/metadata_comparison/metadata_comparison/comparer.py [hunk 9]
```diff
@@ -62,19 +67,23 @@ def compare_jsons(workflow_ids_and_jsons: List[Tuple[str, dict]]) -> pandas.Data
...
-    parser.add_argument('-v', '--verbose', action ='store_true')
-    parser.add_argument('--digester_version', metavar = 'DIGESTERVERSION', type = digester_version_regex_validator, nargs = 1, help = 'Compare digests produced by this version of the digester')
-    parser.add_argument('--digest_gcs_base_path', metavar = 'DIGESTGCSBASEPATH', type = gcs_path_regex_validator, nargs = 1, help = 'GCS base path to the directory containing JSONs produced by digester')
-    parser.add_argument('--output_gcs_file_path', metavar = 'OUTPUTGCSFILE', type = gcs_path_regex_validator, nargs = 1, help = 'GCS path to output CSV file')
-    parser.add_argument('--workflow_ids', metavar = 'WORKFLOWIDS', type = workflow_regex_validator, nargs = '+', help = 'Workflow ids for performance comparison')
+    parser = argparse.ArgumentParser(description='Compare performance metadata JSONs and produce CSV result')
+    parser.add_argument('-v', '--verbose', action='store_true')
+    parser.add_argument('--digester_version', metavar='DIGESTERVERSION', type=digester_version_regex_validator, nargs=1,
```
[the remark this answers is not in the document]
Comment (grsterin) [reply scripts/metadata_comparison/metadata_comparison/comparer.py]: @mcovarr, @cjllanwarne all of a sudden I found out that there are some caveats with kebabs in argparse:
https://stackoverflow.com/a/12834986/1794750
https://stackoverflow.com/a/20250435/1794750

So, I'll have to make amends in this PR

**Pull request merged**